### PR TITLE
Fix #132: Handled Power Enable bit

### LINKED ISSUE #132: Need to improve STM32 power consumption
Sleep power consumption of Catena 4801 is 54 uA and for Catena 4612 is 34 uA with the latest version of BSP and library. The power consumption has to be reduced further.

### PR DESCRIPTION
**Pull request by dhineshkumarmcci**
Cleared and Set the Power enable bit under Stop mode with Power Regulator.

### CODE REVIEW

File: src/lib/stm32/stm32l0/CatenaStm32L0_Sleep.cpp [hunk 1]
```diff
@@ -96,7 +99,7 @@ Name:	CatenaStm32L0::Sleep()
 void CatenaStm32L0::Sleep(uint32_t howLongInSeconds)
 	{
 	this->m_Rtc.SetAlarm(howLongInSeconds);
-	this->m_Rtc.SleepForAlarm(CatenaStm32L0Rtc::SleepMode::Standby);
+	this->m_Rtc.SleepForAlarm(CatenaStm32L0Rtc::SleepMode::StopWithLowPowerRegulator);
```
Comment (chwon64) [inline src/lib/stm32/stm32l0/CatenaStm32L0_Sleep.cpp]: Why did you change?
Did you have power compare data?
According to STM32 BSP doc, standby is most lower power mode.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp [hunk 2]
```diff
@@ -408,6 +411,10 @@ void CatenaStm32L0Rtc::SleepForAlarm(
 		break;
 
 	case SleepMode::StopWithLowPowerRegulator:
+		{
+		/* Set Power interface clock enable bit */
+  		RCC->APB1ENR |= RCC_APB1ENR_PWREN;
```
Comment (chwon64) [inline src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: Why add this code?
When I test this code, I don't have any problem.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp [hunk 3]
```diff
@@ -417,6 +424,29 @@ void CatenaStm32L0Rtc::SleepForAlarm(
 				PWR_SLEEPENTRY_WFI
 				);
 			}
+
+		/* Set Power interface clock enable bit */
+  		RCC->APB1ENR |= RCC_APB1ENR_PWREN;
+
+		if (HAL_RCC_GetHCLKFreq() < 16000000)
```
Comment (chwon64) [inline src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: Same for this code changes. 
As I know, STM32 BSP handles it.

File: src/lib/stm32/stm32l0/CatenaStm32L0_Sleep.cpp (hunk 1 above)
Comment (dhineshkumarmcci) [reply src/lib/stm32/stm32l0/CatenaStm32L0_Sleep.cpp]: Yes ChaeHee, I have compared the power for both modes. We see reduction in power consumption up to 16 uA during **StopWithLowPowerRegulator**.

**NOTE:** Power reduces only with Clear and Set of Power Enable bit under Stop mode. Else we see same power consumption as in Standby mode.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp (hunk 2 above)
Comment (dhineshkumarmcci) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: By handling the Power Enable bit in **StopWithLowPowerRegulator**, we can see reduction in power consumption up to 16 uA.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp (hunk 3 above)
Comment (dhineshkumarmcci) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: By handling the Power Enable bit in **StopWithLowPowerRegulator**, we can see reduction in power consumption up to 16 uA.

Comment (chwon64) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: After wakeup from stop mode, you are setting PWREN bit. I think you should clear it.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp [hunk 4]
```diff
@@ -417,6 +424,29 @@ void CatenaStm32L0Rtc::SleepForAlarm(
...
+			{
+			/* MSI clock, enable and ready */
+			LL_RCC_MSI_Enable();
```
Comment (chwon64) [inline src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: Do you need MSI or HSI clock enable after wakeup from stop mode?

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp (hunk 2 above)
Comment (chwon64) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: All changes in this file should be moved into STM32 BSP HAL_PWR_EnterSTOPMode() in stm32l0xx_hal_pwr.c.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp (hunk 3 above)
Comment (chwon64) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: Did you try to apply PWREN bit with Standby mode?
If you did not try, you should try it to see save more power or not.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp (hunk 2 above)
Comment (dhineshkumarmcci) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: As we get better result with Standby mode, we will move the changes to STM32 BSP under the function HAL_PWR_EnterSTANDBYMode() in stm32l0xx_hal_pwr.c.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp (hunk 3 above)
Comment (dhineshkumarmcci) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: Yes, we have tried out PWREN set and clear in Standby mode. We were able to record 4.2 uA and 24 uA for Catena 4801 and 4612 respectively.

We tried out moving the changes in STM32 BSP _stm32l0xx_hal_pwr.c_ and able to achieve the above result.

File: src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp (hunk 4 above)
Comment (dhineshkumarmcci) [reply src/lib/stm32/stm32l0/CatenaStm32L0Rtc.cpp]: In Stop mode, during a wakeup event, the MSI or HSI clock is selected as system clock depending the bit STOPWUCK in the RCC_CFGR register. 

Due to this reason, we enabled MSI and HSI clock depending on the System clock selected in the IDE. But we do not move these to the STM32 BSP as we made changes only in Standby mode to achieve low power.

**Comment by dhineshkumarmcci** [comment]
@chwon64 We moved all the changes to STM32 BSP **stm32l0xx_hal_pwr.c**, and will be filing a pull request under Arduino_Core_STM32.

As there is no change in Catena-Arduino-Platform now, can we close this pull request?

**Comment by chwon64** [comment]
Changes are moved to STM32 BSP. So close this pull request.
